rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/decompiler/CleanFromMultipleSourceFiles.java

```diff
@@ -0,0 +1,104 @@
+package org.jboss.windup.rules.apps.java.decompiler;
+
+import com.tinkerpop.blueprints.Vertex;
+import com.tinkerpop.gremlin.java.GremlinPipeline;
+import com.tinkerpop.pipes.PipeFunction;
+import org.jboss.windup.config.GraphRewrite;
+import org.jboss.windup.config.operation.GraphOperation;
+import org.jboss.windup.graph.GraphContext;
+import org.jboss.windup.rules.apps.java.model.JavaSourceFileModel;
+import org.ocpsoft.rewrite.context.EvaluationContext;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Removes all the duplicates of .java file within a projectModel. There may be multiple especially in cases where .class is bundled with .java files.
+ * Non-decompiled .java files have higher priority of being deleted than decompiled.
+ */
+public class CleanFromMultipleSourceFiles extends GraphOperation
+{
+    @Override
+    public void perform(GraphRewrite event, EvaluationContext context)
+    {
+        GremlinPipeline<Vertex, Vertex> pipeline = new GremlinPipeline<>(
+                    event.getGraphContext().getQuery().type(JavaSourceFileModel.class).vertices());
+        final GraphContext gContext = event.getGraphContext();
+        pipeline.groupBy(groupByProjectModelFunction(gContext), valueAsFramedVertex(gContext), returnVerticesToDelete(gContext)).cap();
```

Comment:
this groupBy will return a Map containing ProjectModel -> List<JavaSourceFileModel> (for each project model it will list all the JavaSourceFileModel's that should be removed.